src/server/scripts/Northrend/Naxxramas/boss_kelthuzad.cpp

```diff
@@ -45,578 +37,523 @@ enum Texts
-                if (Player* charmed = ObjectAccessor::GetPlayer(*me, itr->first))
-                    charmed->SetObjectScale(itr->second);
+                for (uint8 i = 0; i < nGuardianSpawns; ++i)
+                    _guardianGroups[i] = SUMMON_GROUP_GUARDIAN_FIRST + i;
             }
 
-            chained.clear();
-        }
-
-        void Reset() override
-        {
-            _Reset();
-
-            me->setFaction(35);
-            me->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NON_ATTACKABLE | UNIT_FLAG_REMOVE_CLIENT_CONTROL | UNIT_FLAG_NOT_SELECTABLE);
+            void Reset() override
+            {
+                if (!me->IsAlive())
+                    return;
+                _Reset();
+                me->SetReactState(REACT_PASSIVE);
+                me->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_IMMUNE_TO_PC | UNIT_FLAG_NOT_SELECTABLE);
+                _skeletonCount = 0;
+                _bansheeCount = 0;
+                _abominationCount = 0;
+                _abominationDeathCount = 0;
+                _phaseThree = false;
+            }
+            void EnterEvadeMode(EvadeReason /*why*/) override
+            {
+                for (Data64 portalData : portalList)
+                    if (GameObject* portal = ObjectAccessor::GetGameObject(*me, instance->GetGuidData(portalData)))
+                        portal->SetGoState(GO_STATE_READY);
 
-            ResetPlayerScale();
-            spawns.DespawnAll();
+                Reset();
+                _DespawnAtEvade();
+            }
 
-            instance->SetData(DATA_ABOMINATION_KILLED, 0);
+            void JustSummoned (Creature* summon) override
+            { // prevent DoZoneInCombat
+                summons.Summon(summon);
+            }
 
-            if (GameObject* trigger = ObjectAccessor::GetGameObject(*me, instance->GetGuidData(DATA_KELTHUZAD_TRIGGER)))
+            void KilledUnit(Unit* victim) override
             {
-                trigger->ResetDoorOrButton();
-                trigger->SetPhaseMask(1, true);
+                if (victim->GetTypeId() == TYPEID_PLAYER)
+                    Talk(SAY_SLAY);
             }
 
-            for (uint8 i = 0; i <= 3; ++i)
+            void JustDied(Unit* /*killer*/) override
             {
-                if (GameObject* portal = ObjectAccessor::GetGameObject(*me, instance->GetGuidData(DATA_KELTHUZAD_PORTAL01 + i)))
-                    if (!((portal->getLootState() == GO_READY) || (portal->getLootState() == GO_NOT_READY)))
-                        portal->ResetDoorOrButton();
+                for (ObjectGuid summon : summons)
+                    if (Creature* cSummon = ObjectAccessor::GetCreature(*me, summon))
+                        if (cSummon->IsAlive() && cSummon->GetEntry() == NPC_GUARDIAN)
+                        {
+                            cSummon->AI()->DoAction(ACTION_KELTHUZAD_DIED);
+                            summons.Despawn(cSummon); // prevent them from being despawned by _JustDied
```

Comment:
No. `SummonList::Despawn` is just confusingly named, all it really does is delete the element from map.